gateway-js/src/index.ts

```diff
@@ -255,87 +288,133 @@ export class ApolloGateway implements GraphQLService {
-        this.logger.warn(
-          'Polling Apollo services at a frequency of less than once per 10 seconds (10000) is disallowed. Instead, the minimum allowed pollInterval of 10000 will be used. Please reconfigure your experimental_pollInterval accordingly. If this is problematic for your team, please contact support.',
-        );
-      } else {
-        this.experimental_pollInterval = config.experimental_pollInterval;
-      }
+    return loglevelLogger;
+  }
 
-      // Warn against using the pollInterval and a serviceList simultaneously
-      if (config.experimental_pollInterval && isRemoteConfig(config)) {
-        this.logger.warn(
-          'Polling running services is dangerous and not recommended in production. ' +
-            'Polling should only be used against a registry. ' +
-            'If you are polling running services, use with caution.',
+  private initStaticSchema(config: LocalGatewayConfig | CsdlGatewayConfig) {
+    if (isLocalConfig(config)) {
+      let schema: GraphQLSchema;
+      let composedSdl: string;
+      try {
+        ({ schema, composedSdl } = this.createSchema({
+          serviceList: config.localServiceList,
+        }));
+      } catch (e) {
+        throw Error(
+          "A valid schema couldn't be composed. The following errors were found:\n" +
+            e.message,
         );
       }
 
-      if (config.fetcher) {
-        this.fetcher = config.fetcher;
-      }
+      const queryPlannerPointer = getQueryPlanner(composedSdl);
+      return { schema, composedSdl, queryPlannerPointer };
+    }
+
+    if (isCsdlConfig(config)) {
+      const { schema, composedSdl } = this.createSchema({ csdl: config.csdl });
+      const queryPlannerPointer = getQueryPlanner(config.csdl);
+      return { schema, composedSdl, queryPlannerPointer };
+    }
+
+    // This should be unreachable, and TS even recognizes that `config` is
+    // of type `never` here, but TS still thinks "not all code paths return a value".
+    throw Error("Programming error: impossible config type.");
+  }
+
+  private initQueryPlanStore() {
+    return new InMemoryLRUCache<QueryPlan>({
+      // Create ~about~ a 30MiB InMemoryLRUCache.  This is less than precise
+      // since the technique to calculate the size of a DocumentNode is
+      // only using JSON.stringify on the DocumentNode (and thus doesn't account
+      // for unicode characters, etc.), but it should do a reasonable job at
+      // providing a caching document store for most operations.
+      maxSize:
+        Math.pow(2, 20) *
+        (this.experimental_approximateQueryPlanStoreMiB || 30),
+      sizeCalculator: approximateObjectSize,
+    });
+  }
+
+  private issueDynamicWarningsIfApplicable() {
+    // Warn against a pollInterval of < 10s in managed mode and reset it to 10s
+    if (
+      isManagedConfig(this.config) &&
+      this.config.experimental_pollInterval &&
+      this.config?.experimental_pollInterval < 10000
```

Comment:
I'm surprised that `?` is necessary given the previous line...

Reply:
Yep, it's not
c72b288